0017-negotiation-and-execution-protocol.adoc

```diff
@@ -0,0 +1,544 @@
+
+==== Shortcomings
+
+This first simple negotiation protocol does not really specify an orderbook in the sense that a user can compare ALL available offers.
+A taker can only process the offers of known makers.
+Every taker is responsible for keeping its own list of known makers and evaluate the offers.
+There may exist unknown makers that would offer a better rate.
+
+=== Changes to the existing execution protocol RFC002/RFC003
+
+Constraints:
+
+. Taker defaults to Alice, Maker defaults to Bob
+. The taker has received connection information to contact the maker for execution in the negotiation phase.
+
+==== Input parameters
+
+===== Execution Parameter Definition
+
+For the execution phase as specified by RFC003 we need the parameters for creating the HTLCs on both sides:
+
+* alpha-ledger
+* alpha-asset
+* alpha-amount
+* beta-ledger
+* beta-asset
+* beta-amount
+* expiries
+* hash(secret)
+* identities
+
+In order to come up with these parameters we need:
+
+* role (to know who comes up with the secret)
+
+In order to enable more protocols in the future we need:
+
+* protocol
+
+===== Negotiation Output
+
+From the negotiation phase we have the following output parameters:
+
+[%header,cols=1*]
+|===
+|required
+
+|offer id
+|valid until
+|bid-ledger
+|bid-asset
+|bid-amount
+|ask-ledger
+|ask-asset
+|ask-amount
+|_execution_ protocol
+|_execution_ connection-info (know only to taker)
+|_execution_ role
+|===
+
+[%header,cols=1*]
+|===
+|optional
+
+|_execution_ expiries
+|===
+
+The taker has received the execution connection information from the maker during the negotiation phase. Thus, the taker has to make the first move.
+
+===== Parameter Mapping
+
+This section defines the input parameters for cnd and
+
+[%header,cols=2*]
+|===
+|required (cnd)
+|negotiation
+
+|offer id
+|offer id
+
+|beta-ledger
+|bid-ledger
+
+|beta-asset
+|bid-asset
+
+|beta-amount
+|bid-amount
+
+|alpha-ledger
+|ask-ledger
+
+|alpha-asset
+|ask-asset
+
+|alpha-amount
+|ask-amount
+
+|protocol
+|_execution_ protocol
+
+| connection-info
+|_execution_ connection-info
+
+|role
+|_execution_ role
+
+|===
+
+[%header,cols=2*]
+|===
+|optional (cnd)
+|negotiation
+
+|expiries
+|_execution_ expiries
+
+|identities
+|
+|===
+
+Constraints:
+
+* The connection-info is only provided to the cnd in the role of Alice.
+* If the expiries were not part of the negotiation output cnd will come up with defaults.
+* Identities might be provided by the user. If no identities are provided the cnd will come up with defaults.
+
+==== Communication changes
+
+Instead of a request-response model we propose a messaging model where both parties supply the same parameter to the execution on both sides.
+The CND nodes of the 2 parties execute the swap automatically if the parameters of the counterparty define an executable swap.
+Thus, once one party adds a swap to the pool, cnd sends a message to the counterparty's cnd to verify that the same swap is defined there too.
```

Comment:
What if it is not? Chicken and Egg problem here.
1. Taker informs its cnd
2. Taker cnd asks Maker cnd if same swap is defined
3. Maker cnd says no
4.??? Does the Taker cnd drop the swap?
5. Maker informs its cnd
6. Maker cnd remembers what Taker cnd asks?
7. Maker cnd inform Taker cnd. meaning we always have 2 round of messages?

Reply:
Ok, please clarify that in the document that the taker is the one to send parameters first to confirm.
Which means
Taker:
	1. App tells cnd: here are some parameters
	2. Taker cnd sends message to Maker cnd to confirm swap
	3. Maker cnd waits time X for Maker app to add same swap id (if not done)
	4. Maker cnd replies to Taker cnd

Maker:
	1. App tells cnd: here are some parameters
	2.a (taker cnd already contacted maker cnd): Maker cnd replies to taker cnd to confirm swap
	2.b (taker cnd did not yet contact maker cnd): Maker cnd does nothing and wait for a message from taker cnd.